internal/metadata/ec2/non_imds.go

```diff
@@ -162,3 +165,29 @@ func splitHostname(hostname string) (prefix string, region string, err error) {
 	}
 	return hostname, "", fmt.Errorf("%w: %s", errUnsupportedHostname, hostname)
 }
+
+func fromReservation(reservation ec2.Reservation) (*Metadata, error) {
+	instanceCount := len(reservation.Instances)
+	if instanceCount == 0 || instanceCount > 1 {
+		return nil, fmt.Errorf("%w: %v", errInstanceCount, instanceCount)
+	}
+	instance := reservation.Instances[0]
+	metadata := &Metadata{
+		AccountID:    aws.StringValue(reservation.OwnerId),
+		ImageID:      aws.StringValue(instance.ImageId),
+		InstanceID:   aws.StringValue(instance.InstanceId),
+		InstanceType: aws.StringValue(instance.InstanceType),
+		PrivateIP:    aws.StringValue(instance.PrivateIpAddress),
+	}
+	if instance.Placement != nil {
+		metadata.AvailabilityZone = aws.StringValue(instance.Placement.AvailabilityZone)
+	}
+	return metadata, nil
+}
+
+func getRegionFromAZ(az string) string {
+	if az == "" {
+		return ""
+	}
+	return az[:len(az)-1]
```

Comment:
We hopefully never exhaust a-z from the suffix in a single region :)

Reply:
Based on the current [documentation](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions-availability-zones.html#concepts-availability-zones),
> The code for Availability Zone is its Region code followed by a letter identifier. For example, us-east-1a.

If that ever changes, then this will likely break.